### PR TITLE
fix: Markdown anchor links in package readMe.md broken #2712

### PR DESCRIPTION
**Pull request by coolsp**
Fixes issue #2712 by replacing all href="# occurrences generated by marker with href="${req.headers.referer}#.
Can't be fixed using baseUrl option from marker as it does not allow a base-URL to be added to an URL starting with #.

### CODE REVIEW

**Comment by DanielRuf** [comment]
Hi, thanks for ypur PR.

To me it makes not much sense to use the referer, which is a header that is defined by the browser / client and van be spoofed.

There is probably a safer and better way.

Generally and technically there is no difference between "url#" and "#" if it is displayed on the same hostname / domain.

**Comment by DanielRuf** [comment]
Hi, thanks for your PR.

To me it makes not much sense to use the referer, which is a header that is defined by the browser / client and van be spoofed.

There is probably a safer and better way.

Generally and technically there is no difference between "url#" and "#" if it is displayed on the same hostname / domain.

**Comment by juanpicado** [comment]
Agree with @DanielRuf.

If you want to build `<a href="http://localhost:4873/-/web/detail/@peter/mysample#sample `

You have already the `name` and the url won't change so you might build the href with a function already exist to build the url using the host header (I will share it later on mobile now).

**Comment by coolsp** [comment]
The URL being generated is just used to render an `href` the client user can click (not a link being used by the server directly).  So don't see why a client would spoof, he could just browse to the URL directly.  Also, if this is to prevent a man-in-the-middle attack then I could also change the content, not just the header... (e.g. use TLS to fix).
BUT, what if I omit origin and only use pathname:
```javascript
readMeHtml = readMeHtml.replace(
  /href="#/gi,
  `href="${(new URL(req.headers.referer)).pathname}#`);
```

**Comment by juanpicado** [comment]
We also could rely on the library http://expressjs.com/en/api.html#req.originalUrl instead headers.

**Comment by juanpicado** [comment]
We can discuss the implementation later ( I have some comments ) I just thinking: 
- if Daniel see any sec issue on rely on this header 
- handle scenario where header also could not be there at all (easily removable with a reverse proxy)

**Comment by DanielRuf** [comment]
> handle scenario where header also could not be there at all (easily removable with a reverse proxy)

Same might apply to strict referer policies in browser so no referer or just the domain is available. We need a solution which works also when no referer is provided.

**Comment by coolsp** [comment]
@DanielRuf well today it is 100% broken 24/7 for any condition (I guess since the origin of Verdaccio).  The proposed fix will make it work in +90% of the cases until we find a condition where it doesn't and then we have a use-case we can fix.  I already deployed as a `:local` docker image, works great 😉

**Comment by coolsp** [comment]
I'll did another PR with `pathname` only.

File: src/api/web/endpoint/package.ts [hunk 1]
```diff
@@ -100,7 +100,13 @@ function addPackageWebApi(route: Router, storage: IStorageHandler, auth: IAuth,
         }
 
         res.set(HEADER_TYPE.CONTENT_TYPE, HEADERS.TEXT_PLAIN);
-        next(parseReadme(info.name, info.readme));
+        let readMeHtml = parseReadme(info.name, info.readme);
```
Comment (juanpicado) [inline src/api/web/endpoint/package.ts:103]: I'd prefer to have this option within the readme package https://github.com/verdaccio/monorepo/blob/9.x/core/readme/src/index.ts
sorry if is in another repo, but at this time 5.x still depends of monorepo (master has already another `readme` package but not usable on v5).

eg:
```
function parseReadme(readme: string, options = {}): string | void {```
  // current logic

  ...if(options.pathname){ 
   // apply your patch
  } 
}
```
I think in this way is the option is closest to the reason of change, we can test (please 🙏🏼  add some tests) it easily and we won't need to mutate `readmeHtml` on the line `105`. Once we release on `monorepo ` we can just update the version on this PR.

Then instead `if ('string' === typeof readMeHtml && 'string' === typeof req.headers.referer) {` you could just don't include or not the `pathname` as option.

File: src/api/web/endpoint/package.ts [hunk 2]
```diff
@@ -100,7 +100,13 @@ function addPackageWebApi(route: Router, storage: IStorageHandler, auth: IAuth,
...
+        if ('string' === typeof readMeHtml && 'string' === typeof req.headers.referer) {
+          readMeHtml = readMeHtml.replace(
+            /href="#/gi,
+            `href="${(new URL(req.headers.referer)).pathname}#`);
```
Comment (juanpicado) [inline src/api/web/endpoint/package.ts:107]: Instead `req.headers.referer` there is a safest way using `request.get()` http://expressjs.com/en/api.html#req.get

File: src/api/web/endpoint/package.ts (hunk 1 above)
Comment (coolsp) [reply src/api/web/endpoint/package.ts:103]: @juanpicado done

**Comment by juanpicado** [comment]
Hi @coolsp looks good, but there are some conflicts we need to resolve first.

**Comment by juanpicado** [comment]
If you need some support or help ping me, I could help with yarn.

**Comment by coolsp** [comment]
@juanpicado I rebased on PR #2727

**Comment by juanpicado** [comment]
great, could you copy the PR body to the new PR, I'd like to keep tracking why the change.

**Comment by coolsp** [comment]
Oops... done !

**Pull request closed without merging**